### PR TITLE
feat: more informative error message on xml validation errors

### LINKED ISSUE #37788: [Issue] feat: more informative error message on xml validation errors
This issue is automatically created based on existing pull request: magento/magento2#36836: feat: more informative error message on xml validation errors

---------
### Description (*)
This PR adds more informative output to the existing error message displayed, when there is invalid XML configuration.

For example if there is invalid layout.xml the standard error message displayed to a developer would be something like:
```
1 exception(s):
Exception #0 (Magento\Framework\Config\Dom\ValidationException): Element 'arguments': This element is not expected.
Line: 1471
```
This isn't very helpful as the line number refers to the XML after it has been merged, making it pretty impossible to determine the file that caused the issue.
This PR changes the output to
```
1 exception(s):
Exception #0 (Magento\Framework\Config\Dom\ValidationException): Element 'arguments': This element is not expected.
Line: 1471
The xml was: 
1466:            <argument name="show_category" xsi:type="boolean">true</argument>
1467:         </arguments>
1468:      </block>
1469:      <referenceBlock name="product_view_detail" remove="true"/>
1470:      <block name="abc.gtm.product_view_detail" class="ABC\GoogleTagManager\Block\Detail" template="ABC_GoogleTagManager::detailproduct.phtml"/>
1471:      <arguments>
1472:         <argument name="show_category" xsi:type="boolean">true</argument>
1473:      </arguments>
1474:   </referenceContainer>
1475:</body>
```
Which although the line numbers still refer to the merged XML, it will allow the developer to search the source files for a more specific string. Making it much easier to find the source of the issue.

### Related Pull Requests
https://github.com/magento-gl/magento2ee/pull/205

### Fixed Issues (if relevant)
N/A

### Manual testing scenarios (*)
1. Put the site into development mode, and disable cache.
2. Add a breaking change to a layout XML file. e.g add: `<bar>foo</bar>`
3. Load the frontend of the site,

You should get an error message that you can use to find the offending configuration.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by adamzero1**
### Description (*)
This PR adds more informative output to the existing error message displayed, when there is invalid XML configuration.

For example if there is invalid layout.xml the standard error message displayed to a developer would be something like:
```
1 exception(s):
Exception #0 (Magento\Framework\Config\Dom\ValidationException): Element 'arguments': This element is not expected.
Line: 1471
```
This isn't very helpful as the line number refers to the XML after it has been merged, making it pretty impossible to determine the file that caused the issue.
This PR changes the output to
```
1 exception(s):
Exception #0 (Magento\Framework\Config\Dom\ValidationException): Element 'arguments': This element is not expected.
Line: 1471
The xml was: 
1466:            <argument name="show_category" xsi:type="boolean">true</argument>
1467:         </arguments>
1468:      </block>
1469:      <referenceBlock name="product_view_detail" remove="true"/>
1470:      <block name="abc.gtm.product_view_detail" class="ABC\GoogleTagManager\Block\Detail" template="ABC_GoogleTagManager::detailproduct.phtml"/>
1471:      <arguments>
1472:         <argument name="show_category" xsi:type="boolean">true</argument>
1473:      </arguments>
1474:   </referenceContainer>
1475:</body>
```
Which although the line numbers still refer to the merged XML, it will allow the developer to search the source files for a more specific string. Making it much easier to find the source of the issue.

### Related Pull Requests
https://github.com/magento-gl/magento2ee/pull/205

### Fixed Issues (if relevant)
N/A

### Manual testing scenarios (*)
1. Put the site into development mode, and disable cache.
2. Add a breaking change to a layout XML file. e.g add: `<bar>foo</bar>`
3. Load the frontend of the site,

You should get an error message that you can use to find the offending configuration.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#37788: feat: more informative error message on xml validation errors

### CODE REVIEW

**Comment by hostep** [comment]
Nice!

However, there was already another improvement made in this area recently I believe: https://github.com/magento/magento2/pull/36239 (it looks like this one is not going to be included in Magento 2.4.6, because it's not in 2.4.6-beta5, and that would really suck, so we'll have to wait until the year 2024 for this one to be included in Magento 2.4.7 ...)

Anyway, not sure if we need both improvements?

File: lib/internal/Magento/Framework/Config/Dom.php [hunk 1]
```diff
@@ -398,10 +399,11 @@ public static function validateDomDocument(
      *
      * @param \LibXMLError $errorInfo
      * @param string $format
+     * @param \DOMDocument|null $dom
      * @return string
      * @throws \InvalidArgumentException
      */
-    private static function _renderErrorMessage(\LibXMLError $errorInfo, $format)
+    private static function _renderErrorMessage(\LibXMLError $errorInfo, $format, $dom = null)
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Config/Dom.php]:
```suggestion
    private static function _renderErrorMessage(\LibXMLError $errorInfo, string $format, \DOMDocument $dom = null): string
```

File: lib/internal/Magento/Framework/Config/Dom.php [hunk 2]
```diff
@@ -424,6 +426,14 @@ private static function _renderErrorMessage(\LibXMLError $errorInfo, $format)
...
             }
         }
+        if ($dom) {
+            $xml = explode(PHP_EOL, $dom->saveXml());
+            $lines = array_slice($xml, max(0, $errorInfo->line - 5), 10, true);
+            $result .= 'The xml was: '.PHP_EOL;
+            foreach ($lines as $lineNumber => $line) {
+                $result .= $lineNumber.':'.$line.PHP_EOL;
+            }
+        }
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Config/Dom.php]:
```suggestion
        if ($dom) {
            $xml = explode(PHP_EOL, $dom->saveXml());
            $lines = array_slice($xml, max(0, $errorInfo->line - 5), 10, true);
            $result .= 'The xml was: ' . PHP_EOL;
            foreach ($lines as $lineNumber => $line) {
                $result .= $lineNumber . ':' . $line . PHP_EOL;
            }
        }
```

File: lib/internal/Magento/Framework/Config/Dom.php [hunk 3]
```diff
@@ -380,7 +381,7 @@ public static function validateDomDocument(
         try {
             $result = $dom->schemaValidate($schema);
             if (!$result) {
-                $errors = self::getXmlErrors($errorFormat);
+                $errors = self::getXmlErrors($errorFormat, $dom);
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Config/Dom.php:385]: Can we pass the `$dom` object in all the places where we're calling the `getXmlErrors` method?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Could you also please cover your change with any type of automated tests?

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Config/Dom.php (hunk 3 above)
Comment (adamzero1) [reply lib/internal/Magento/Framework/Config/Dom.php:385]: was unable to replicate the scenario where `!$result` and didn't want to add somewhere I couldn't test.
But yes, I can't see arason why not.

**Comment by adamzero1** [comment]
> Nice!
> 
> However, there was already another improvement made in this area recently I believe: #36239 (it looks like this one is not going to be included in Magento 2.4.6, because it's not in 2.4.6-beta5, and that would really suck, so we'll have to wait until the year 2024 for this one to be included in Magento 2.4.7 ...)
> 
> Anyway, not sure if we need both improvements?

I'm not sure of the performance implications of merging on everyfile, considering the number of files that are processed.
Though this would help the file be identified.
I will have a look to see if there is a way to get best of both

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Could you also please cover your change with any type of automated tests and fix failing tests?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Unit Tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magentoa run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
It seems some automated tests are failing due to PR changes. Hence moving this to `Changes Requested`

**Comment by engcom-Echo** [comment]
@magento run Unit Tests,Integration Tests,Functional Tests EE,Functional Tests CE,Functional Tests B2B

**Review by engcom-Hotel — APPROVED** [review]
It seems the failed tests are flaky. Hence approving this PR.

**Comment by engcom-Lima** [comment]
@magento create issue

**Comment by engcom-Lima** [comment]
@magento run Functional Tests EE , Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Lima** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`** and **`PHP 8.1`**

**Manual testing scenario:**

1. Put the site into development mode, and disable the cache.
2. Add a breaking change to a layout XML file. e.g add: <bar>foo</bar>
3. Load the front end of the site.

**Before:** :heavy_multiplication_x: Complete information was not available of error message on XML validation errors.
Screenshot 2023-07-24 at 12 19 48 PM
Screenshot 2023-07-24 at 12 24 52 PM




**After:** :heavy_check_mark:  Now complete information is getting displayed on XML validation errors.
Screenshot 2023-07-24 at 12 26 59 PM
Screenshot 2023-07-24 at 12 27 24 PM
Screenshot 2023-07-24 at 12 27 32 PM

Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Echo** [comment]
@magento run Unit Tests

**Comment by engcom-Echo** [comment]
@magento run Unit Tests

**Comment by engcom-Echo** [comment]
Functional Failure are not related to the PR changes.
Hence moving to `merge in progress`.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run WebAPI Tests,Unit Tests,Integration Tests,Functional Tests EE

**Comment by engcom-Echo** [comment]
Functional Tests EE failures are different on last two run on same commit. Other build failures are not related to PR.
Functional Tests EE 
Run1
36836-ee

Run2
Screenshot 2023-08-18 at 12 31 27 PM

**Pull request merged**